File: src/org/symbolmath/exception/SyntaxException.java
package org.symbolmath.exception;

public class SyntaxException extends ParseException {

  public SyntaxException(String message, String text, int index) {
    super(getError(message, text, index));
  }

  private static String getError(String message, String text, int index) {
    return message + "\n" + text + "\n" + spaces(index) + "^";
  }

  private static String spaces(int n) {
    StringBuilder f = new StringBuilder(n);
    for (int i = 0; i < n; i++) {
      f.append(' ');
    }
    return f.toString();
  }

}

File: src/main/java/org/symbolmath/exception/SyntaxException.kt
package org.symbolmath.exception

class SyntaxException(message: String, text: String, index: Int) : ParseException(getError(message, text, index)) {
    companion object {
        private fun getError(message: String, text: String, index: Int): String {
            return """
                $message
                $text
                ${spaces(index)}^
                """.trimIndent()
        }

        private fun spaces(n: Int): String {
            return " ".repeat(Math.max(0, n))
        }
    }
}